File: my-folder/0496-next-greater-element-i/solution.ts
class Stacks {
    private stack: number[]
    constructor() {
        this.stack = []
    }
    push(data: number) {
        this.stack.push(data)
    }
    pop(): number {
        if (this.isEmpty())
            throw new Error("Index out of bounds")
        return this.stack.pop()!
    }
    peak(): number {
        return this.stack[this.stack.length-1]
    }
    length(): number {
        return this.stack.length
    }
    isEmpty(): boolean {
        return this.length() == 0
    }
}

function nextGreaterElement(nums1: number[], nums2: number[]): number[] {
    let result: number[] = []
    let stack = new Stacks()
    let mapping = new Map<number, number>()

    for (let i = 0; i < nums2.length; i++) {
        if (stack.isEmpty()) {
            stack.push(nums2[i])
        } else if (stack.peak() >= nums2[i]) {
            stack.push(nums2[i])
        }
        // Recursion
        greaterNum(stack, nums2[i], mapping)
    }
    return nums1.map((v, i) => {
        return mapping.get(v) || -1
    })
}

function greaterNum(stack: Stacks, v: number, map: Map<number, number>) {
    if (stack.peak() < v) {
        let popped = stack.pop()
        map.set(popped, v)
        greaterNum(stack, v, map)
    }
    stack.push(v)
}

File: my-folder/0496-next-greater-element-i/solution.go
type Stack []int

func (s *Stack) Push(v int) {
  *s = append(*s, v)
}

func (s *Stack) Pop() (int, bool) {
	if (*s).IsEmpty() {
        return 0, false
	}
	index := len(*s) - 1
	element := (*s)[index]
	*s = (*s)[:index]
	return element, true
}

func (s *Stack) IsEmpty() bool {
  return len(*s) == 0
}

func (s *Stack) Peak() (int, bool) {
	if (*s).IsEmpty() {
        return 0, false
	}

	index := len(*s) - 1
	return (*s)[index], true
}

func nextGreaterElement(nums1 []int, nums2 []int) []int {
    var stack Stack
    maps := make(map[int]int)
    result := make([]int, len(nums1))

    for _, v := range nums2 {
        for !stack.IsEmpty() {
            peak, _ := stack.Peak()
            if peak < v {
                top, _ := stack.Pop()
                maps[top] = v
            } else {
                break
            }
		}
        stack.Push(v)
    }

    for i, v := range nums1 {
        if val, ok := maps[v]; ok {
            result[i] = val
        } else {
            result[i] = -1
        }
    }

    return result
}
